README.md
# bmi734_hw1

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2019 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2019-02-09
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2019-02-09
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/bmi734_hw1/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/bmi734_hw1/compare/0.1.0...0.1.1


test/bmi734_hw1/core_test.clj
(ns bmi734-hw1.core-test
  (:require [clojure.test :refer :all]
            [bmi734-hw1.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/bmi734_hw1/problem-1.clj
(ns bmi734-hw1.problem-1
  (:gen-class)
  (:import (org.opencv.core Mat Core CvType)
           (org.opencv.imgproc Imgproc)
           (org.opencv.imgcodecs Imgcodecs)))
(defn -main []
  (write-cleaned-image)
  (write-segmented-image)
  )

(defn write-cleaned-image []
                               (let [embryos (Imgcodecs/imread "resources/embryos.jpg")
                                     embryos_cleaned (Mat.)]
                                (Imgproc/medianBlur embryos embryos_cleaned 3)
                                (Imgcodecs/imwrite "embryos_cleaned.jpg" embryos_cleaned)
                               )
  )
(defn write-segmented-image []
                                 (def mri-brain (Imgcodecs/imread "resources/mri_brain.jpg"))
                                   (def mri-brain-segmented (Mat.))
                                   (Imgproc/Canny mri-brain mri-brain-segmented 75 200)
                                 (let [
                                     imsize (.size mri-brain-segmented)
                                     zeros  (Mat/zeros imsize CvType/CV_8U)
                                     overlay (Mat.)
                                     output (Mat.)
                                     ]
                                 (Core/merge [zeros zeros mri-brain-segmented] overlay)
                                 (Core/scaleAdd overlay 255 mri-brain output)
                                 (Imgcodecs/imwrite "mri_brain_segd.jpg" output)
                                 )
                               )
  (defn generate-test-images [input-filename output-filename]
    (let [input-image (Imgcodecs/imread input-filename)
          ;steps (range 4096 65536 (/ 65536 16))
          steps (range 70 80 1)
          neg (Mat.)
          ;steps (range 100 200 5)
          ;steps (range 200 300 5)
          ;steps (range 300 400 5)
          output-image (Mat.)]
      (doseq [step steps]
        (println step)
        (Core/bitwise_not input-image neg)
        (Imgproc/Canny neg output-image 200 step)
        (Imgcodecs/imwrite (str "tmp/" output-filename step ".jpg") output-image)
        )
      )
    )

src/bmi734_hw1/problem-2.clj
(ns bmi734-hw1.problem-2
(:gen-class)
  (:require [bmi734-hw1.OpenCVHelpers :as OpenCVHelpers])
  (:import (org.opencv.core Mat Core CvType Scalar)
           (org.opencv.imgcodecs Imgcodecs)
           (org.opencv.imgproc Imgproc)
          ))
(defn get-segmented-image [] (
                               (def mri-brain (Imgcodecs/imread "resources/mri_brain.jpg"))
                               (def mri-brain-segmented (Mat.))
                               (Imgproc/Canny mri-brain mri-brain-segmented 75 200)
                               (let [imsize (.size mri-brain-segmented)
                                     b (Mat. imsize CvType/CV_8U)
                                     g (Mat. imsize CvType/CV_8U)
                                     r (Mat. imsize CvType/CV_8U)
                                     zeros  (Mat/zeros imsize CvType/CV_8U)
                                     overlay (Mat.)
                                     output (Mat.)
                                     normaled (Mat.)]
                                 (Core/merge [zeros zeros mri-brain-segmented] overlay)
                                 (Core/scaleAdd overlay 255 mri-brain output)
                                 (OpenCVHelpers/display-mat output)
                                 )
                               )
(defn generate-test-images [input-filename output-filename]
    (let [input-image (Imgcodecs/imread input-filename)
         ;steps (range 4096 65536 (/ 65536 16))
          steps (range 70 80 1)
          neg (Mat.)
          ;steps (range 100 200 5)
          ;steps (range 200 300 5)
          ;steps (range 300 400 5)
         output-image (Mat.)]
         (doseq [step steps]
           (println step)
           (Core/bitwise_not input-image neg)
           (Imgproc/Canny neg output-image 200 step)
           (Imgcodecs/imwrite (str "tmp/" output-filename step ".jpg") output-image)
         )
    )
)

src/bmi734_hw1/OpenCVHelpers.clj
(ns bmi734-hw1.OpenCVHelpers
  (:gen-class)
  (:import
           (org.opencv.highgui HighGui)
           (javax.swing ImageIcon JLabel JFrame)
           (java.awt FlowLayout)))
(defn display-mat [mat]
  ;OpenCV's HighGUI doesn't work very well so I adapted this nice example:
  ;http://answers.opencv.org/question/8119/how-to-display-image-on-java-release/
  (let [frame (HighGui/createJFrame (:name (meta mat)) HighGui/WINDOW_AUTOSIZE)
        buffered-img (HighGui/toBufferedImage mat)
        icon (ImageIcon. buffered-img)
        h (+ 50 (.getHeight buffered-img nil))
        w (+ 50 (.getWidth buffered-img nil))
        lbl (JLabel.)
        ]
    (.setLayout frame (FlowLayout.))
    (.setSize frame w h)
    (.setIcon lbl icon)
    (.add frame lbl)
    (.setVisible frame true)
    (.setDefaultCloseOperation frame JFrame/DISPOSE_ON_CLOSE)
    )
  )

src/bmi734_hw1/problem_1.clj
(ns bmi734-hw1.problem-1
  (:gen-class)
  (:import (org.opencv.core Mat Core CvType)
           (org.opencv.imgproc Imgproc)
           (org.opencv.imgcodecs Imgcodecs)))

(defn write-cleaned-image []
                               (let [embryos (Imgcodecs/imread "resources/embryos.jpg")
                                     embryos_cleaned (Mat.)]
                                (Imgproc/medianBlur embryos embryos_cleaned 3)
                                (Imgcodecs/imwrite "embryos_cleaned.jpg" embryos_cleaned)
                               )
  )
(defn write-segmented-image []
                                 (def mri-brain (Imgcodecs/imread "resources/mri_brain.jpg"))
                                   (def mri-brain-segmented (Mat.))
                                   (Imgproc/Canny mri-brain mri-brain-segmented 75 200)
                                 (let [
                                     imsize (.size mri-brain-segmented)
                                     zeros  (Mat/zeros imsize CvType/CV_8U)
                                     overlay (Mat.)
                                     output (Mat.)
                                     ]
                                 (Core/merge [zeros zeros mri-brain-segmented] overlay)
                                 (Core/scaleAdd overlay 255 mri-brain output)
                                 (Imgcodecs/imwrite "mri_brain_segd.jpg" output)
                                 )
                               )
  (defn generate-test-images [input-filename output-filename]
    (let [input-image (Imgcodecs/imread input-filename)
          ;steps (range 4096 65536 (/ 65536 16))
          steps (range 70 80 1)
          neg (Mat.)
          ;steps (range 100 200 5)
          ;steps (range 200 300 5)
          ;steps (range 300 400 5)
          output-image (Mat.)]
      (doseq [step steps]
        (println step)
        (Core/bitwise_not input-image neg)
        (Imgproc/Canny neg output-image 200 step)
        (Imgcodecs/imwrite (str "tmp/" output-filename step ".jpg") output-image)
        )
      )
    )

(defn -main []
  (write-cleaned-image)
  (write-segmented-image)
  )
